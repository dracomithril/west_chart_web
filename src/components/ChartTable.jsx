/**
 * Created by Gryzli on 26.01.2017.
 */
import React from 'react';
import PropTypes from 'prop-types';
import ReactTable from 'react-table';
import { Checkbox, Image, Label, OverlayTrigger, Tooltip } from 'react-bootstrap';
import { getFbPictureUrl } from '../utils/utils';
import './components.css';

import action_types from './../reducers/action_types';

function formatDate(date) {
  const date2 = new Date(date);
  const yearNow = new Date().getFullYear();
  const options = {
    day: 'numeric',
    month: 'short',
    hour: '2-digit',
    minute: '2-digit',
    second: '2-digit',
  };
  if (yearNow === date2.getFullYear()) {
    return date2.toLocaleString('pl-PL', options);
  }
  return date2.toLocaleString('pl-PL', { year: 'numeric', ...options });
}

const getTime = date => new Date(date).getTime();
const ChartTable = ({ data }, { store }) => {
  const { user, show_wait } = store.getState();
  const TimeCell = ({ value: time_value }) => (
    <div>
      <span style={{ color: 'red' }}>c: {formatDate(time_value.created_time)}</span>
      <br />
      {time_value.created_time !== time_value.updated_time && (
        <span style={{ color: 'green' }}>u: {formatDate(time_value.updated_time)}</span>
      )}
    </div>
  );

  TimeCell.propTypes = {
    value: PropTypes.object,
  };
  const time = {
    Header: <i className="fa fa-clock-o">Time</i>,
    id: 'createTime',
    resizable: true,
    minWidth: 150,
    maxWidth: 200,
    accessor: d => ({
      created_time: getTime(d.created_time),
      updated_time: getTime(d.updated_time),
    }),
    Cell: TimeCell,
  };
  const videoLinkCell = ({ value }) =>
    value.url === undefined ? (
      <span>{value.title}</span>
    ) : (
      <a href={value.url} target="_newtab">
        {value.title}
      </a>
    );
  videoLinkCell.propTypes = {
    value: PropTypes.object,
  };
  const link = {
    Header: (
      <i className="fa fa-external-link" style={{ color: 'red' }} aria-hidden="true">
        Link
      </i>
    ), // Custom header components!
    accessor: d => d.link,
    minWidth: 200,
    width: 300,
    maxWidth: 400,
    id: 'yt_link',
    Cell: videoLinkCell,
  };
  const UserInfoCell = ({ value: from }) => (
    <div style={{ textAlign: 'left' }}>
      <Image style={{ float: 'left' }} src={getFbPictureUrl(from.id)} />
      <div style={{ display: 'inline-grid' }}>
        <span style={{ paddingLeft: 10 }}>{from.first_name}</span>
        <span style={{ paddingLeft: 10 }}>{from.last_name}</span>
      </div>
    </div>
  );
  UserInfoCell.propTypes = {
    value: PropTypes.object,
  };
  const ReactionCell = ({ value }) => {
    const condition = value && value.message;
    return (
      <div>
        {condition && (
          <OverlayTrigger
            placement="bottom"
            overlay={<Tooltip id="tooltip">{value.message}</Tooltip>}
          >
            <Label bsStyle="success">
              <i className="fa fa-envelope-o" aria-hidden="true" />
            </Label>
          </OverlayTrigger>
        )}
        {!condition && (
          <Label bsStyle="danger">
            <i className="fa fa-times" aria-hidden="true" />
          </Label>
        )}
        <div>
          <i className="fa fa-thumbs-o-up" style={{ color: 'blue' }} aria-hidden="true" />
          {value.reactions_num}
        </div>
      </div>
    );
  };
  ReactionCell.propTypes = {
    value: PropTypes.object,
  };
  const post_info = {
    Header: 'Post Info',
    columns: [
      {
        Header: (
          <i className="fa fa-user-circle" style={{ color: 'green' }} aria-hidden="true">
            user
          </i>
        ),
        resizable: true,
        minWidth: 140,
        maxWidth: 180,
        id: 'user',
        accessor: 'from', // String-based value accessors !
        Cell: UserInfoCell,
      },
      {
        Header: <i className="fa fa-info-circle" style={{ color: 'blue' }} aria-hidden="true" />,
        accessor: ({ message, reactions_num }) => ({ message, reactions_num }),
        id: 'woc_f',
        maxWidth: 50,
        Cell: ReactionCell,
      },
      time,
      link,
    ],
  };

  const SelectCell = ({ value, row }) => (
    <Checkbox
      bsClass="checkbox1"
      checked={value}
      id={row.id}
      name="selected"
      onChange={({ target }) => {
        store.dispatch({
          type: 'TOGGLE_SELECTED',
          id: target.id,
          checked: target.checked,
        });
      }}
    />
  );
  SelectCell.propTypes = {
    value: PropTypes.object,
    row: PropTypes.object,
  };
  const columns = [
    {
      Header: 'id',
      show: false,
      accessor: 'id',
    },
    {
      Header: <small>{data.length}</small>,
      columns: [
        {
          sortable: false,
          resizable: false,
          Header: () => (
            <Checkbox
              bsClass="checkbox1"
              onClick={() => {
                store.dispatch({ type: action_types.TOGGLE_ALL });
              }}
            />
          ),
          width: 40,
          accessor: 'selected',
          Cell: SelectCell,
        },
      ],
    },
    post_info,
  ];
  const tableOptions = {
    filterable: false,
    // filtered:[{id:'woc_f'},{id:'user'}],
    resizable: false,
    pageSizeOptions: [20, 50, 100],
  };
  return (
    <ReactTable
      data={data}
      className="-striped -highlight"
      {...tableOptions}
      columns={columns}
      defaultPageSize={20}
      minRows={10}
      loading={show_wait}
      noDataText={
        <span>
          {`Hi, ${user.first_name} please click `}
          <strong style={{ color: 'blue' }}>Update</strong>
          {` to start.`}
        </span>
      }
    />
  );
};

ChartTable.contextTypes = {
  store: PropTypes.object,
};
ChartTable.propTypes = {
  data: PropTypes.array,
};
export default ChartTable;
