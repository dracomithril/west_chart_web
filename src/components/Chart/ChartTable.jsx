/**
 * Created by Gryzli on 26.01.2017.
 */
import React from 'react';
import PropTypes from 'prop-types';
import ReactTable from 'react-table';
import 'react-table/react-table.css';
import { Checkbox, Image, OverlayTrigger, Tooltip } from 'react-bootstrap';
import { getFbPictureUrl } from '../../utils/utils';
import { action_types } from './../../reducers/action_types';
import '../components.css';
import '../../App.css';

function formatDate(date) {
  const date2 = new Date(date);
  const yearNow = new Date().getFullYear();
  const options = {
    day: 'numeric',
    month: 'numeric',
  };
  if (yearNow === date2.getFullYear()) {
    return date2.toLocaleString('pl-PL', options);
  }
  return date2.toLocaleString('pl-PL', { year: 'numeric', ...options });
}

function fullFormatDate(date) {
  const date2 = new Date(date);
  const yearNow = new Date().getFullYear();
  const options = {
    day: 'numeric',
    month: 'numeric',
    hour: '2-digit',
    minute: '2-digit',
    second: '2-digit',
  };
  if (yearNow === date2.getFullYear()) {
    return date2.toLocaleString('pl-PL', options);
  }
  return date2.toLocaleString('pl-PL', { year: 'numeric', ...options });
}

const getTime = date => new Date(date).getTime();
const ChartTable = ({ data }, { store }) => {
  const { user, show_wait } = store.getState();
  const TimeCell = ({ value: time_value }) => (
    <div>
      <span
        style={{ color: 'red' }}
        title={`creation time: ${fullFormatDate(time_value.created_time)}`}
      >
        <i className="fas fa-caret-right" /> {formatDate(time_value.created_time)}
      </span>
      <br />
      {time_value.created_time !== time_value.updated_time && (
        <span
          style={{ color: 'green' }}
          title={`update time: ${fullFormatDate(time_value.updated_time)}`}
        >
          <i className="fas fa-caret-up" /> {formatDate(time_value.updated_time)}
        </span>
      )}
    </div>
  );

  TimeCell.propTypes = {
    value: PropTypes.object,
  };
  const time = {
    Header: <i className="far fa-clock">Time</i>,
    id: 'createTime',
    resizable: true,
    minWidth: 100,
    maxWidth: 200,
    accessor: d => ({
      created_time: getTime(d.created_time),
      updated_time: getTime(d.updated_time),
    }),
    Cell: TimeCell,
  };
  const videoLinkCell = ({ value }) =>
    value.url === undefined ? (
      <span>{value.title}</span>
    ) : (
      <a href={value.url} target="_newtab">
        {value.title}
      </a>
    );
  videoLinkCell.propTypes = {
    value: PropTypes.object,
  };
  const link = {
    Header: (
      <i className="fas fa-external-link-square-alt" style={{ color: 'red' }} aria-hidden="true">
        Link
      </i>
    ), // Custom header components!
    accessor: d => d.link,
    minWidth: 200,
    width: 300,
    maxWidth: 400,
    id: 'yt_link',
    Cell: videoLinkCell,
  };
  const UserInfoCell = ({ value: from }) => (
    <div style={{ textAlign: 'left' }}>
      <Image style={{ float: 'left' }} src={getFbPictureUrl(from.id)} />
      <div style={{ display: 'inline-grid' }}>
        <span style={{ paddingLeft: 10 }}>{from.first_name || from.name.split(' ')[0]}</span>
        <span style={{ paddingLeft: 10 }}>{from.last_name || from.name.split(' ')[1]}</span>
      </div>
    </div>
  );
  UserInfoCell.propTypes = {
    value: PropTypes.object,
  };
  const ReactionCell = ({ value }) => {
    const condition = value && value.message;
    return (
      <div>
        {condition ? (
          <OverlayTrigger
            placement="bottom"
            overlay={<Tooltip id="tooltip">{value.message}</Tooltip>}
          >
            <i
              className="fas fa-comment"
              style={{ color: 'green', cursor: 'pointer' }}
              aria-hidden="true"
            />
          </OverlayTrigger>
        ) : (
          <i className="fas fa-times-circle" style={{ color: 'red', cursor: 'no-drop' }} />
        )}
        <div>
          <i
            className="far fa-thumbs-up"
            style={{ color: 'blue' }}
            aria-hidden="true"
            title="people reactions count"
          />
          {value.reactions_num}
        </div>
      </div>
    );
  };
  ReactionCell.propTypes = {
    value: PropTypes.object,
  };
  const post_info = [
    {
      Header: (
        <i className="fas fa-user-circle" style={{ color: 'green' }} aria-hidden="true">
          user
        </i>
      ),
      resizable: true,
      minWidth: 140,
      maxWidth: 180,
      id: 'user',
      accessor: 'from',
      Cell: UserInfoCell,
    },
    {
      Header: <i className="fas fa-info-circle" style={{ color: 'blue' }} aria-hidden="true" />,
      accessor: ({ message, reactions_num }) => ({ message, reactions_num }),
      id: 'woc_f',
      maxWidth: 50,
      Cell: ReactionCell,
    },
    time,
    link,
  ];

  const SelectCell = ({ value, row, onChange }) => (
    <Checkbox
      bsClass="chart-table__checkbox"
      checked={value}
      id={row.id}
      name="selected"
      onChange={({ target }) => {
        store.dispatch({
          type: 'TOGGLE_SELECTED',
          id: target.id,
          checked: target.checked,
        });
      }}
    />
  );
  SelectCell.propTypes = {
    value: PropTypes.object,
    row: PropTypes.object,
    onChange: PropTypes.func,
  };
  const columns = [
    {
      Header: 'id',
      show: false,
      accessor: 'id',
    },

    {
      sortable: false,
      resizable: false,
      Header: () => (
        <Checkbox
          bsClass="chart-table__checkbox"
          onClick={() => {
            store.dispatch({ type: action_types.TOGGLE_ALL });
          }}
        />
      ),
      width: 40,
      accessor: 'selected',
      Cell: SelectCell,
    },
    ...post_info,
  ];
  const tableOptions = {
    filterable: false,
    // filtered:[{id:'woc_f'},{id:'user'}],
    resizable: false,
    pageSizeOptions: [20, 50, 100],
  };
  return (
    <ReactTable
      data={data}
      className="-striped -highlight"
      {...tableOptions}
      columns={columns}
      defaultPageSize={20}
      minRows={10}
      loading={show_wait}
      noDataText={
        <span>
          {`Hi, ${user.first_name} please click `}
          <strong style={{ color: 'blue' }}>Update</strong>
          {` to start.`}
        </span>
      }
    />
  );
};

ChartTable.contextTypes = {
  store: PropTypes.object,
};
ChartTable.propTypes = {
  data: PropTypes.array,
};
export default ChartTable;
