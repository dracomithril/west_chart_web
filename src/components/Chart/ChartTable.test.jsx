import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { shallowToJson } from 'enzyme-to-json';
import { ChartTable } from './ChartTable';

Enzyme.configure({ adapter: new Adapter() });
const { chart: data } = require('../../___data___/response.json');


describe('<ChartTable/>', () => {
  it('renders without crashing', () => {
    const wrapper = shallow(
      <ChartTable data={[]} />,
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
  it('render with objects', () => {
    const partData = data.slice(0, 5);
    const wrapper = shallow(
      <ChartTable data={partData} />,
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
