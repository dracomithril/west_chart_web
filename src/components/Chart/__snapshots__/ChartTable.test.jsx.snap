// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChartTable/> render with objects 1`] = `
<div>
  <div
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
      }
    }
  >
    <WithStyles(WithFormControlContext(FormControlLabel))
      control={
        <WithStyles(Checkbox)
          color="primary"
          id="selectAll"
          onChange={[Function]}
          value="selectAll"
        />
      }
      label="select all"
    />
  </div>
  <ChartRow
    checked={true}
    createdTime="2018-03-16T00:12:32+0000"
    id="1707149242852457_2150864795147564"
    key="1707149242852457_2150864795147564"
    link={
      Object {
        "name": "EZA - Shape of You",
        "title": "EZA - Shape of You",
        "type": "share",
        "url": "https://www.youtube.com/watch?v=ALIBUPY_UEc",
      }
    }
    message="Ship of a black madness, trade the power! "
    onChange={[Function]}
    reactionsNum={4}
    updatedTime="2018-03-16T00:16:21+0000"
  />
  <ChartRow
    checked={false}
    createdTime="2018-03-16T00:12:59+0000"
    id="1707149242852457_2139869112913799"
    key="1707149242852457_2139869112913799"
    link={
      Object {
        "name": "Sam Smith - Too Good At Goodbyes (Sofia Karlberg Cover)",
        "title": "Sam Smith - Too Good At Goodbyes (Sofia Karlberg Cover)",
        "type": "share",
        "url": "https://www.youtube.com/watch?v=rLKreFwTSYs",
      }
    }
    message={null}
    onChange={[Function]}
    reactionsNum={0}
    updatedTime="2018-03-16T00:12:59+0000"
  />
  <ChartRow
    checked={false}
    createdTime="2018-03-16T00:12:39+0000"
    id="1707149242852457_2151609491739761"
    key="1707149242852457_2151609491739761"
    link={
      Object {
        "name": "Iggy Azalea - Savior (Lyric Video) ft. Quavo",
        "title": "Iggy Azalea - Savior (Lyric Video) ft. Quavo",
        "type": "share",
        "url": "https://youtu.be/LxoGNkLpgZk",
      }
    }
    message="https://youtu.be/LxoGNkLpgZk"
    onChange={[Function]}
    reactionsNum={2}
    updatedTime="2018-03-16T00:12:40+0000"
  />
  <ChartRow
    checked={false}
    createdTime="2018-03-16T00:12:36+0000"
    id="1707149242852457_2150975431803167"
    key="1707149242852457_2150975431803167"
    link={
      Object {
        "name": "Calum Scott - Only You (Audio)",
        "title": "Calum Scott - Only You (Audio)",
        "type": "share",
        "url": "https://www.youtube.com/watch?v=WX_0qCyFaeM",
      }
    }
    message="https://www.youtube.com/watch?v=WX_0qCyFaeM"
    onChange={[Function]}
    reactionsNum={3}
    updatedTime="2018-03-16T00:12:36+0000"
  />
  <ChartRow
    checked={false}
    createdTime="2018-03-14T23:10:55+0000"
    id="1707149242852457_2149545078612869"
    key="1707149242852457_2149545078612869"
    link={
      Object {
        "name": "Dolly Parton - Nine to Five",
        "title": "Dolly Parton - Nine to Five",
        "type": "share",
        "url": "https://www.youtube.com/watch?v=LwDMFOLIHxU",
      }
    }
    message="https://www.youtube.com/watch?v=LwDMFOLIHxU"
    onChange={[Function]}
    reactionsNum={1}
    updatedTime="2018-03-14T23:47:48+0000"
  />
</div>
`;

exports[`<ChartTable/> renders without crashing 1`] = `
<div
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "justifyContent": "center",
      "minHeight": 300,
      "textAlign": "center",
    }
  }
>
  No data
</div>
`;
